be/src/olap/olap_server.cpp

```diff
@@ -391,4 +313,70 @@ void StorageEngine::_tablet_checkpoint_callback(DataDir* data_dir) {
     } while (!_stop_background_threads_latch.wait_for(MonoDelta::FromSeconds(interval)));
 }
 
-}  // namespace doris
+void StorageEngine::_compaction_tasks_producer_callback() {
+#ifdef GOOGLE_PROFILER
+    ProfilerRegisterThread();
+#endif
+    LOG(INFO) << "try to start compaction producer process!";
+
+    std::vector<DataDir*> data_dirs;
+    for (auto& tmp_store : _store_map) {
+        data_dirs.push_back(tmp_store.second);
+    }
+
+    int round = 0;
+    CompactionType compaction_type;
+    do {
+        if (!config::disable_auto_compaction) {
+            if (round < config::cumulative_compaction_rounds_for_each_base_compaction_round) {
```

Comment:
And I also test this with following case:
1. Only 1 BE with 1 data dir.
2. Create one table with 100 buckets.
3. insert data into this table for every 5 seconds.

The compaction is triggered every 2 seconds. And each compaction task cost just 0.x seconds. But the average version count of tablets is about 50, and can not be lower.

So I think the way to generate compaction tasks through polling may not be appropriate. One possible way is to generate compaction tasks through triggering.

Based on polling, currently only one task can be done in 2 seconds, and based on triggering, in my case, it can be done 500 times per second (because the amount of data in each batch is very small in the case of high-frequency load)

Reply:
Thanks for your suggestions! I optimized the implementation logic of my producer.  If all the compaction tasks produced can hold `permits`, the producer will continue to produce compaction tasks without sleep. In this way, the production speed can far meet consumer demand.